text/projects/Krzysztof_Michalik/Wee_LCP.tex

```diff
@@ -0,0 +1,179 @@
+\end{enumerate}
+
+Rozważmy analogiczny podział każdego 'krótkiego' ($< \kappa^2$) przedziału na takie przedziały (zwane dalej miniprzedziałami), że ostatnim elementem każdego miniprzedziału będzie bit '1' oraz każdy z miniprzedziałów będzie zawierał $\lambda = \lfloor \log{\kappa}^2 \rfloor$ bitów '1' (być może poza ostatnim). Oznacza to, ze utworzymy w ten sposób co najwyżej $\frac{n}{\lambda}$ miniprzedziałów (sumując po wszystkich miniprzedziałach utworzonych ze wszystkich 'krótkich' przedziałów).
+
+Analogicznie będziemy chcieli zapamiętać co najwyżej 2 wartości dla każdego miniprzedziału zawierającego się w przedziale $P$:
+\begin{enumerate}
+    \item $N'[i] = select_1(P, i\lambda)$.
+    \item Jeśli miniprzedział jest dłuższy niż $\log{n}^{\delta}$, czyli $\log{n}^{\delta} > N'[i] - N'[i-1]$, to zhadrcodujemy odpowiedzi na wszystkie możliwe zapytania $select_1()$ na tym miniprzedziale w tablicy $P'$ i zapiszemy wskaźnik do niej.
+    \item Jeśli przedział jest krótszy niż $\log{n}^{\delta}$, to nie musimy nic zapisywać.
+\end{enumerate}
+
+%Teraz zhardcodujemy odpowiedzi na pytania $select_1(S, i\kappa)$ dla przedziałów długości $\kappa ^2$. Zapisane wartości będą rzędu $O(\log{n})$ dla każdego z długich przedziałów. Sprawia to, że zhardcodowane wartości zajmą $O(\frac{n\log{n}}{\kappa}) = O(\frac{n}{\log{n}})$ miejsca.
+
+%W dalszej kolejności zhardcodujemy jeszcze kilka tablic, każda asymptotycznie nie zajmująca więcej miejsca od powyższej.
+
+%Analogicznie postąpimy dla 'krótkich' przedziałów, dalej nazywanych miniblokami. Mowa tu o przedziałach które są mniejsze niż $\kappa^2$.
+
+%Niech $\lambda = \lfloor \log{\kappa}^2 \rfloor = \Theta(\log^2{\log{n}})$. Jak wyżej dzieliliśmy bloki na długie i krótkie, teraz dzielimy tak minibloki. Każdy miniblok będzie zawierał $\lambda$ bitów '1'. Następnie dla każdego bloku $>\log{n}^{\delta}$ hadrcodujemy odpowiedzi na zapytania.
+
+Teraz zastanówmy się jak możemy odpowiadać na zapytania typu $select_1(S, i)$. Mamy kilka dość intuicyjnych przypadków:
+
+\begin{enumerate}
+    \item Długość przedziału w którym jest $i$-ta jedynka ($N[\lfloor \frac{i}{\kappa} \rfloor + 1] - N[\lfloor \frac{i}{\kappa} \rfloor]$) jest dłuższa niż $\kappa^2$ - wtedy nasza odpowiedź jest zhadrcodowana i wystarczy ją odczytać w czasie $O(1)$.
+    \item Długość przedziału jest mniejsza niż $\kappa^2$, ale długość miniprzedziału jest większa od $\log{n}^{\delta}$. Wtedy odpowiedź również jest zhardcodowana, jako suma dwóch wartości. Odczytujemy je również w czasie $O(1)$.
+    \item Długość miniprzedziału jest mniejsza niż $\log{n}^{\delta}$. Wtedy nasza wartość nie jest zhardcodowana, ale możemy popatrzyć się na początek miniprzedziału i podać wartość dla niego jako naszą wstępną odpowiedź (wiemy, że aktualna odpowiedź na zapytanie $select_1()$ będzie nie mniejsza). Dokładną wartość poznajemy porównując w czasie $O(\log{n}^{\delta})$ fragmenty $T$.
```

Comment:
"Dokładną wartość poznajemy porównując w czasie $O(\log{n}^{\delta})$ fragmenty $T$." - does it mean that we just traverse T from the given position and add 1 whenever we encounter 1 in T?